patching/EAP7-1063_EAP_Patching_With_Galleon.adoc

```diff
@@ -0,0 +1,121 @@
+
+* feature specs and feature groups (units of configuration);
+
+* resources (various resources included into feature-pack, including module artifact versions);
+
+* provisioning plugin (it is possible to patch a provisioning plugin included in the original feature-pack).
+
+Patch in this proposal is always assumed to be a one-off patch. There is no notion of a cumulative patch. A CP is assumed to be a version update in Galleon mechanism.
+
+==== Atomic provisioning state transition
+
+Whenever a patch or a version update is applied (which can happen simultaneously for multiple projects/products in the installation as an atomic state transition) the actual distribution is going to be modified only in case the new configuration could be successfully provisioned. For that reason the updated version is always provisioned at a temporary staged location. Only after the updated version has been successfully built, the installation moves from the staged location to the original home location.
+
+==== Provisioning state history
+
+Galleon saves a description (in terms of feature-pack configurations) of the current provisioning state before every provisioning operation (such as install, uninstall, provision state, update) into the state history which allows the user to undo the last operation go back to the previous state (undo requires access to the artifact repository as well). A user can limit or disable the state history.
+
+==== User changes and provisioning state transition
+
+Before installing a new feature-pack (patch, version update, new software) into existing installation, Galleon will identify user changes since the last provisioned state. That includes changes to the configuration and other files. The changes are calculated as a diff between the last provisioned state and the current state.
+
+Next, Galleon provisions an updated installation at a staged location and if that went ok, it applies the user changes identified previously to the newly provisioned state.
+
+User changes to the files that haven't changed in the updated installation are re-applied to the updated installation.
+
+If a user changed a file in the installation that has changed in the updated installation then if the changes can be merged (text files, property files) they are merged. If the changes can't be merged, the file with the user changes will appear next to the updated file with a suffix ".prev".
+
+Standalone configuration changes are identified as diffs in terms of feature configurations and then re-applied to the corresponding updated configurations.
+
+Host and domain configuration changes are not currently merged as the standalone ones and *not* merged as text files. If user modified host or domain configurations and the corresponding configurations have changed in the updated installation, the user-modified configuration files will appear in the installation with suffix ".prev".
```

Comment:
This is a regression versus current patching. Current patching ignores the config files, so the user changes are preserved.

Reply:
It can be viewed as a regression. On the other hand, keeping the previous strategy will be wrong in the new implementation simply because the new mechanism allows the user to re-configure the installation, extend it by including additional modules or installing new products, reduce it by excluding installed modules or uninstalling products, etc. Each of those operations may significantly affect the current configuration. Keeping the current config as the active one in this case may result in a broken state.

Reply:
I don't mean to advocate keeping the previous strategy; sorry if I was unclear about that. The concern is dropping the user modifications; how/why they didn't get dropped before isn't important.

The proposed strategy for standalone is fine. My concern is not applying a similar strategy to domain/host means user modifications get dropped (i.e. stored to .prev, but not in the primary file.)

Reply:
I share your concern. All we need to do to support the diff and merging is to export domain and host configs as a list of ops or preferably features. And I guess that's doable.

Reply:
What I think we need here is read-resource(recursive) that would return features. Doing the diff we are doing for standalone currently is too expensive (feature-diff operation). If we had that read-features operation that would have been enough and easier to do for domain as well.

Reply:
Actually, more like describe-model but in features.

Reply:
Might be better to have .new files as opposed to .prev.
See also http://people.ds.cam.ac.uk/jw35/docs/rpm_config.html

Note that 3 way merges have been tried in the past, but so far I'm not aware of great success.